### PR TITLE
Matt/user scopes

### PR DESCRIPTION
**Pull request by MattKersey**
Automatically assign OAuth scopes to users based on their stores and roles within the stores. Each store has three scopes - employee, manager, and vendor. Employees are assigned only the employee scope, managers get both the employee and manager scopes, and vendors get all three. I have protected our endpoints based on these scopes as follows:

```
GET /api/users/  -  any authenticated user
GET /api/users/<pk>/  -  any authenticated user
PUT /api/users/<pk>/  -  any authenticated user
POST /api/users/<pk>/add_store/  -  any authenticated user
POST /api/users/<pk>/remove_store/  -  users with manager scope
DELETE /api/users/<pk>/delete_store/  -  users with vendor scope
GET /api/users/current_user/  -  any authenticated user
GET /api/stores/  -  any authenticated user
GET /api/stores/<pk>/  -  users with employee scope
PUT /api/stores/<pk>/  -  users with vendor scope
POST /api/stores/<pk>/purchase_items/  -  users with employee scope
POST /api/stores/<pk>/add_item/  -  users with vendor scope
POST /api/stores/<pk>/remove_item/  -  users with vendor scope
DELETE /api/stores/delete_item/  -  users with vendor scope
GET /api/items/  -  any authenticated user
GET /api/items/<pk>/  -  any authenticated user
POST /api/items/  -  users with vendor scope
PUT /api/items/<pk>/  -  users with vendor scope
```

### CODE REVIEW

File: backend/backend/scopes.py [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+            scopes.append("store_" + str(store.pk) + ":vendor")
+        return scopes
+
+    def get_default_scopes(self, application=None, request=None, *args, **kwargs):
+        scopes = ["read", "write"]
+        for store in Store.objects.all():
+            scopes.append("store_" + str(store.pk) + ":employee")
+            scopes.append("store_" + str(store.pk) + ":manager")
+            scopes.append("store_" + str(store.pk) + ":vendor")
+        return scopes
```
Comment (ceguiluzrosas) [inline backend/backend/scopes.py:39]: Will these methods be different in the future? @MattKersey

File: backend/backend/scopes.py [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+        except ImproperlyConfigured:
+            required_scopes = []
+        pk = ""
+        if request.data.get("store_id"):
+            pk = str(request.data.get("store_id"))
+        elif getPK(request.path) != "":
+            pk = getPK(request.path)
+        else:
+            return required_scopes
+        # TODO: Add in method switches
```
Comment (ceguiluzrosas) [inline backend/backend/scopes.py:55]: 1. Sorta like the same question above: Will all these classes have the same get_scopes method in the future?

2. Can you clarify the methods switches comment?

File: backend/api/tests/test_views.py [hunk 3]
```diff
@@ -29,12 +29,17 @@ def setupOAuth(self):
         skip_authorization=True,
     )
     self.application.save()
+    scope = "read write"
+    for pk in stores:
+        scope += " store_" + str(pk) + ":vendor"
+        scope += " store_" + str(pk) + ":manager"
+        scope += " store_" + str(pk) + ":employee"
```
Comment (ceguiluzrosas) [inline backend/api/tests/test_views.py:36]: @MattKersey Just to confirm. For these test cases, we are giving the `test-user` all scopes (vendor, manager, and employee) of permissions only for the stores we defined above, right? 

If that's the case, should we also include some test cases where someone without the proper scopes tries to do an out-of-scope request?

**Comment by ceguiluzrosas** [comment]
Looks good!! 🔥  👍🏽

**Review by ceguiluzrosas — CHANGES_REQUESTED** [review]
Comments above.

File: backend/backend/scopes.py (hunk 1 above)
Comment (MattKersey) [reply backend/backend/scopes.py:39]: Not that I can tell, this structure is just following what I found in the oauth toolkit docs, which has identical methods for the two. I can simplify and call a common method.

**Review by ceguiluzrosas — APPROVED** [review]
Nice changes 🥳  LGTM 👍🏽

**Pull request merged**